File: threads/src/main/java/im/threads/internal/helpers/FileHelper.java
package im.threads.internal.helpers;

import android.content.Context;

import java.io.File;

import im.threads.business.logger.LoggerEdna;
import im.threads.business.transport.models.AttachmentSettings;
import im.threads.business.config.BaseConfig;
import im.threads.internal.chat_updates.ChatUpdateProcessor;
import im.threads.internal.utils.PrefUtils;

public enum FileHelper {

    INSTANCE;

    private static final double MEGABYTE = 1024 * 1024;

    FileHelper() {
        ChatUpdateProcessor.getInstance().getAttachmentSettingsProcessor()
                .subscribe(attachmentSettings -> saveAttachmentSettings(attachmentSettings.getContent()),
                        LoggerEdna::error
                );
    }

    public static File createImageFile(Context context) {
        String filename = "thr" + System.currentTimeMillis() + ".jpg";
        File output = new File(context.getFilesDir(), filename);
        LoggerEdna.debug("File genereated into filesDir : " + output.getAbsolutePath());
        return output;
    }

    public boolean isAllowedFileSize(long fileSize) {
        return fileSize / MEGABYTE <= getMaxAllowedFileSize();
    }

    public boolean isAllowedFileExtension(String fileExtension) {
        AttachmentSettings.Content attachmentSettings = getAttachmentSettings();
        for (String allowedExt : attachmentSettings.getFileExtensions()) {
            if (allowedExt.equalsIgnoreCase(fileExtension)) {
                return true;
            }
        }
        return false;
    }

    public long getMaxAllowedFileSize() {
        return getAttachmentSettings().getMaxSize();
    }

    private void saveAttachmentSettings(AttachmentSettings.Content attachmentSettingsContent) {
        PrefUtils.setAttachmentSettings(BaseConfig.instance.gson.toJson(attachmentSettingsContent));
    }

    private AttachmentSettings.Content getAttachmentSettings() {
        String settingsStr = PrefUtils.getAttachmentSettings();
        if (settingsStr.isEmpty()) {
            return getDefaultAttachmentSettings();
        } else {
            AttachmentSettings.Content attachmentSettingsContent
                    = BaseConfig.instance.gson.fromJson(settingsStr, AttachmentSettings.Content.class);
            if (attachmentSettingsContent == null) {
                return getDefaultAttachmentSettings();
            } else {
                return attachmentSettingsContent;
            }
        }
    }

    private AttachmentSettings.Content getDefaultAttachmentSettings() {
        return new AttachmentSettings.Content(30, new String[]{"jpeg", "jpg", "png", "pdf", "doc", "docx", "rtf"});
    }
}

File: src/main/java/im/threads/ui/utils/FileHelper.kt
package im.threads.ui.utils

import android.annotation.SuppressLint
import im.threads.business.chatUpdates.ChatUpdateProcessor
import im.threads.business.config.BaseConfig
import im.threads.business.logger.LoggerEdna
import im.threads.business.preferences.Preferences
import im.threads.business.serviceLocator.core.inject
import im.threads.business.transport.models.AttachmentSettings
import im.threads.ui.preferences.PreferencesUiKeys
import io.reactivex.disposables.Disposable

/**
 * Вспомогательный класс для работы с файлами
 */
@SuppressLint("CheckResult")
object FileHelper {
    private val chatUpdateProcessor: ChatUpdateProcessor by inject()
    private val preferences: Preferences by inject()

    private const val MEGABYTE = (1024 * 1024).toDouble()
    private var disposable: Disposable? = null

    init {
        subscribeToAttachments()
    }

    fun subscribeToAttachments() {
        if (disposable == null ||
            disposable?.isDisposed == true ||
            !chatUpdateProcessor.attachmentSettingsProcessor.hasSubscribers()
        ) {
            disposable = chatUpdateProcessor.attachmentSettingsProcessor
                .subscribe(
                    { receivedAttachmentSettings ->
                        LoggerEdna.error("subscribeToAttachments(). receivedAttachmentSettings: $receivedAttachmentSettings")
                        receivedAttachmentSettings.content?.let { attachmentSettings = it }
                    },
                    LoggerEdna::error
                )
        }
    }

    fun isAllowedFileSize(fileSize: Long): Boolean {
        return fileSize / MEGABYTE <= maxAllowedFileSize
    }

    fun isAllowedFileExtension(fileExtension: String?): Boolean {
        attachmentSettings?.let {
            for (allowedExt in it.fileExtensions) {
                if (allowedExt.equals(fileExtension, ignoreCase = true)) {
                    return true
                }
            }
        }
        LoggerEdna.warning("isAllowedFileExtension() - false. Extension \"$fileExtension\" missing in array: ${attachmentSettings?.fileExtensions?.asList()}")
        return false
    }

    fun isFileExtensionsEmpty(): Boolean {
        LoggerEdna.error("isFileExtensionsEmpty(???). Extension array: ${attachmentSettings?.fileExtensions?.asList()}")
        return attachmentSettings?.fileExtensions?.isEmpty() ?: true
    }

    fun isJpgAllow(): Boolean {
        return attachmentSettings?.fileExtensions?.contains("jpg") ?: false
    }

    val maxAllowedFileSize: Long
        get() = attachmentSettings?.maxSize?.toLong() ?: 0

    private var attachmentSettings: AttachmentSettings.Content?
        get() {
            val settingsStr = preferences.get(PreferencesUiKeys.PREF_ATTACHMENT_SETTINGS) ?: ""
            return BaseConfig.getInstance().gson.fromJson(settingsStr, AttachmentSettings.Content::class.java)
        }
        set(value) = preferences.save(
            PreferencesUiKeys.PREF_ATTACHMENT_SETTINGS,
            BaseConfig.getInstance().gson.toJson(value)
        )
}
